### ISSUE TITLE
Custom ceph.conf for cephcsi pods

### ISSUE THREAD

**Issue by Madhu-1** [rook/rook#9220]
Currently, the admin/user can configure the `ceph.conf` for daemon pods using https://rook.io/docs/rook/v1.7/ceph-advanced-configuration.html#custom-cephconf-settings. This issue is to have a custom `ceph.conf` for the clients/cephcsi pods. the support for the same is added in cephcsi https://github.com/ceph/ceph-csi/pull/2476. 

Below are the points I discussed with @travisn 
* Add new `csi-ceph-conf.yaml` to the example directory (this is a new configmap as this ceph.conf is applicable for only clients)
* If the configmap is created update cephcsi templates to mount the configmap to csi pods
* If the configmap is not created cephcsi will use the default `ceph.conf` what it creates
*  Update documentation for the new configmap

**Note:-** Even though the cephcsi/Rook operator can handle multiple ceph clusters. As cephcsi is installed only once per rook operator we can have only once `ceph.conf` for csi pods for all the ceph clusters.

Below is the little clarity am looking for. ( didn't get a chance to think about it before discussing it with Travis)

* Do we need to add owner-ref for this configmap or user will take care of deleting it manually as part of the cleanup process?

* How to handle the upgrade process. if cephcsi makes something detailed we just need to document the new changes so the user can update the `ceph.conf` he created?

@travisn please add if I missed anything from the discussion

cc @travisn @leseb @BlaineEXE

**Comment by travisn** [comment]
What about the following?
- The Rook operator always creates the new csi configmap (with the default config) if it does not exist. If the configmap already exists, rook will not create or update it.
- When Rook starts the CSI driver, it will always mount the configmap to the csi pods. (No need to make this optional if rook just always guarantees the configmap exists)

Upgrades will be handled already by this process to convert from the hard-coded config that was used by csi, to the configurable option. Before this change, there is currently no way for the user to customize the settings, correct?

**Comment by leseb** [comment]
One question, why don't we have a new parameter to the storage class that would represent the name of the ConfigMap that contains extra settings to apply? I assume that you don't need all the details right? There is really a minimum config you need like the cluster FSID and the mon IP. All the rest can be considered as extra, these extra could be fetched from the CM. Or a new entry in the SC could represent the ceph.conf, right?
Today the SC has some Ceph config flags exposed like `imageFormat` and `imageFeatures` which could be actually ceph flags.

If we have the setting in the SC, this could be much more granular and also allow a different config per ceph cluster right?
Even different configs within the same cluster.

**Comment by travisn** [comment]
> If we have the setting in the SC, this could be much more granular and also allow a different config per ceph cluster right?
Even different configs within the same cluster.

There is one csi driver per operator, so the csi driver cannot have separate settings per cluster. 
If we were to support the configmap to be defined in the storageclass, then seems like every ceph command executed by the csi driver would need to specify the ceph.conf requested by that storageclass. I assume that would be a big change for the csi driver so we would need a clear scenario for the feature.

**Comment by leseb** [comment]
> > If we have the setting in the SC, this could be much more granular and also allow a different config per ceph cluster right?
> > Even different configs within the same cluster.
> 
> There is one csi driver per operator, so the csi driver cannot have separate settings per cluster.

I don't understand this, why not? If the driver reads the config it is given it should work regardless if they is one or more cluster, right?

**Comment by travisn** [comment]
> > > If we have the setting in the SC, this could be much more granular and also allow a different config per ceph cluster right?
> > > Even different configs within the same cluster.
> > 
> > 
> > There is one csi driver per operator, so the csi driver cannot have separate settings per cluster.
> 
> I don't understand this, why not? If the driver reads the config it is given it should work regardless if there is one or more clusters, right?

@Madhu-1 Can comment on this for sure, but I believe the CSI driver only has a single ceph.conf, and the proposal with this issue is just to make that customizable. If we were to support a different ceph.conf per storage class, the csi driver would need to be much smarter about which ceph.conf is being used depending on the cluster being targeted, instead of ceph just picking up the default ceph.conf.

**Comment by leseb** [comment]
> > > > If we have the setting in the SC, this could be much more granular and also allow a different config per ceph cluster right?
> > > > Even different configs within the same cluster.
> > > 
> > > 
> > > There is one csi driver per operator, so the csi driver cannot have separate settings per cluster.
> > 
> > 
> > I don't understand this, why not? If the driver reads the config it is given it should work regardless if there is one or more clusters, right?
> 
> @Madhu-1 Can comment on this for sure, but I believe the CSI driver only has a single ceph.conf, and the proposal with this issue is just to make that customizable. If we were to support a different ceph.conf per storage class, the csi driver would need to be much smarter about which ceph.conf is being used depending on the cluster being targeted, instead of ceph just picking up the default ceph.conf.

I discussed this with @Madhu-1 yesterday. We agreed to stick with the current proposal but it will be nice to have a per-SC configuration possible too. That would allow us to handle scenarios like multi-cluster and multiple configs for the same cluster. The biggest downside now is that any config changes will affect all the SCs which might not be desired. Especially if one SC is being debugged or something. The main challenge would be on how to populate different configurations and dynamically load them.
One thing we should investigate is the possibility to use the mon config store...
